tools/relay-emulator/main.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/http"
+	"os"
+	"slices"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/primev/mev-commit/x/util"
+	"github.com/urfave/cli/v2"
+)
+
+var (
+	optionL1RPCURL = &cli.StringFlag{
+		Name:    "l1-rpc-url",
+		Usage:   "URL of the L1 RPC server",
+		EnvVars: []string{"MOCK_RELAY_L1_RPC_URL"},
+	}
+
+	optionHTTPPort = &cli.IntFlag{
+		Name:    "http-port",
+		Usage:   "port to listen on for HTTP requests",
+		EnvVars: []string{"MOCK_RELAY_HTTP_PORT"},
+		Value:   8080,
+	}
+
+	optionLogFmt = &cli.StringFlag{
+		Name:    "log-fmt",
+		Usage:   "log format to use, options are 'text' or 'json'",
+		EnvVars: []string{"MOCK_RELAY_LOG_FMT"},
+		Value:   "text",
+		Action: func(ctx *cli.Context, s string) error {
+			if !slices.Contains([]string{"text", "json"}, s) {
+				return fmt.Errorf("invalid log-fmt, expecting 'text' or 'json'")
+			}
+			return nil
+		},
+	}
+
+	optionLogLevel = &cli.StringFlag{
+		Name:    "log-level",
+		Usage:   "log level to use, options are 'debug', 'info', 'warn', 'error'",
+		EnvVars: []string{"MOCK_RELAY_LOG_LEVEL"},
+		Value:   "info",
+		Action: func(ctx *cli.Context, s string) error {
+			if !slices.Contains([]string{"debug", "info", "warn", "error"}, s) {
+				return fmt.Errorf("invalid log-level, expecting 'debug', 'info', 'warn', 'error'")
+			}
+			return nil
+		},
+	}
+
+	optionLogTags = &cli.StringFlag{
+		Name:    "log-tags",
+		Usage:   "log tags is a comma-separated list of <name:value> pairs that will be inserted into each log line",
+		EnvVars: []string{"MOCK_RELAY_LOG_TAGS"},
+		Action: func(ctx *cli.Context, s string) error {
+			for i, p := range strings.Split(s, ",") {
+				if len(strings.Split(p, ":")) != 2 {
+					return fmt.Errorf("invalid log-tags at index %d, expecting <name:value>", i)
+				}
+			}
+			return nil
+		},
+	}
+)
+
+func main() {
+	app := &cli.App{
+		Name:  "relay-emulator",
+		Usage: "Emulates the required relay APIs",
+		Flags: []cli.Flag{
+			optionL1RPCURL,
+			optionHTTPPort,
+			optionLogFmt,
+			optionLogLevel,
+			optionLogTags,
+		},
+		Action: func(c *cli.Context) error {
```

Comment:
There is no graceful shutdown for the HTTP server. Consider using its `Shutdown` method with a combination of `c.Context` and `signal.NotifyContext`.